Add bintime_subx() macro

---
 sys/time.h |   11 +++++++++++
 1 file changed, 11 insertions(+)

Index: sys/sys/time.h
===================================================================
--- sys.orig/sys/time.h	2011-06-28 19:37:17.000000000 +1000
+++ sys/sys/time.h	2011-06-29 13:17:46.000000000 +1000
@@ -72,20 +72,31 @@ bintime_add(struct bintime *bt, const st
 	uint64_t u;
 
 	u = bt->frac;
 	bt->frac += bt2->frac;
 	if (u > bt->frac)
 		bt->sec++;
 	bt->sec += bt2->sec;
 }
 
 static __inline void
+bintime_subx(struct bintime *bt, uint64_t x)
+{
+	uint64_t u;
+
+	u = bt->frac;
+	bt->frac -= x;
+	if (u < bt->frac)
+		bt->sec--;
+}
+
+static __inline void
 bintime_sub(struct bintime *bt, const struct bintime *bt2)
 {
 	uint64_t u;
 
 	u = bt->frac;
 	bt->frac -= bt2->frac;
 	if (u < bt->frac)
 		bt->sec--;
 	bt->sec -= bt2->sec;
 }
